Ejh ncid issue test clean up and port to yellowstone

jedwards4b opened the pull request:
My update of Ed's branch, now passes all tests with NTASKS = 4 on yellowstone.  Tests with NTASKS > 4 still have some issues.  The test test_iosystem3_simple2 appears to pass but then exits with a non-zero code.

edhartnett commented:
Let's comment out the tests that are failing and do the merge, then I will go back and get those tests working again.

The failing tests are all subsets of what pio_iosystems_test2 and pio_iosystems_test3 do, so as long as those two tests pass, the functionality is there.

jedwards4b commented:
I think that commenting out or removing tests was, in part, what lead us into the current mess.  I have now addressed the remaining issues with tests on this branch and rebased to the current master.   Please review.

Katetc commented on tests/unit/test_iosystem2_simple2.c:
```diff
@@ -119,6 +122,8 @@ main(int argc, char **argv)
     if ((ret = PIOc_finalize(iosysid_world)))
 	ERR(ret);
+    }  /* my_rank < TARGET_NTASKS */
+    MPI_Barrier(MPI_COMM_WORLD);
```
Why is this MPI_COMM_WORLD and not test_comm?

Katetc commented on tests/unit/test_iosystem3.c:
```diff
@@ -305,37 +307,39 @@ main(int argc, char **argv)
+	  ERR(ret);
+
+    } /* my_rank < TARGET_NTASKS */
+    MPI_Barrier(MPI_COMM_WORLD);
```
Again, why MPI_COMM_WORLD and not test_comm? Also, is this barrier really needed before finalize? If so, maybe it should be moved into the finalize function to ensure that it is always there when needed?

jedwards4b commented:
This MPI_COMM_WORLD prevents the tasks which are not in test_comm from
calling finalize before the test is completed.

On Tue, Oct 18, 2016 at 10:35 AM, Kate Thayer-Calder <
notifications@github.com> wrote:

> *@Katetc* commented on this pull request.
> ------------------------------
>
> In tests/unit/test_iosystem3.c
> <https://github.com/NCAR/ParallelIO/pull/122#pullrequestreview-4709699>:
>
> >      	if ((ret = MPI_Comm_free(&even_comm)))
> -    	    ERR(ret);
> +	  ERR(ret);
> +
> +    } /* my_rank < TARGET_NTASKS */
> +    MPI_Barrier(MPI_COMM_WORLD);
>
> Again, why MPI_COMM_WORLD and not test_comm? Also, is this barrier really
> needed before finalize? If so, maybe it should be moved into the finalize
> function to ensure that it is always there when needed?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/ParallelIO/pull/122#pullrequestreview-4709699>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF16GG9kSBmN8gxUPhY_JTqiPjctDh8Gks5q1PVrgaJpZM4KZWbB>
> .
>



-- 
Jim Edwards

CESM Software Engineer
National Center for Atmospheric Research
Boulder, CO

jayeshkrishna commented on src/clib/pio_internal.h:
```diff
@@ -76,8 +76,10 @@ extern "C" {
+    int pio_num_iosystem(int *niosysid);
 
     file_desc_t *pio_get_file_from_id(int ncid);
+    int pio_get_file_from_id2(int ncid, file_desc_t **filep);
```
How is pio_get_file_from_id2() different from pio_get_file_from_id() and is the new function used in the code?

edhartnett replied:
It is not really different and I will remove it.

I thought there was a bug in pio_get_gile_from_id() so I added pio_get_file_from_id2() as a total rewrite.

But that was not where the bug was, so I will remove it.

jayeshkrishna commented on tests/general/CMakeLists.txt:
```diff
@@ -317,10 +317,10 @@ if (PIO_USE_MPISERIAL)
-        EXECUTABLE ${CMAKE_CURRENT_BINARY_DIR}/pio_decomp_fillval
-        NUMPROCS 4
-        TIMEOUT ${DEFAULT_TEST_TIMEOUT})
+    # add_mpi_test(pio_decomp_fillval
```
This test was re-enabled on master. It should be enabled on this branch too.

jedwards4b replied:
thanks missed that.

jedwards4b replied:
All other tests pass but I haven't figured this one out yet.

jayeshkrishna commented on tests/unit/test_common.c:
```diff
@@ -43,6 +43,38 @@
+    num++;
+    format[fmtidx++] = PIO_IOTYPE_NETCDF;
+#endif
+#ifdef _NETCDF4
```
Doesn't _NETCDF4 also imply PIO_IOTYPE_NETCDF?

jedwards4b replied:
No _NETCDF4 implies only NETCDF4P and NETCDF4C

jayeshkrishna replied:
Hmmm, the fortran testing framework assumes PIO_IOTYPE_NETCDF when _NETCDF4 is enabled (I always assumed if you have _NETCDF4 it means you also have _NETCDF. Is there a case where _NETCDF4 is enabled by not _NETCDF? Or does _NETCDF4 imply that _NETCDF is also enabled?). I will fix that.

jedwards4b replied:
If you have _NETCDF4 you do have _NETCDF but not the reverse

jayeshkrishna replied:
ok, so the fortran testing framework code is also correct. Thanks.

jedwards4b commented on tests/unit/test_async_simple.c:
```diff
@@ -90,11 +91,11 @@ main(int argc, char **argv)
-    } /* endif comp_task */
-
+      } /* endif comp_task */
+    } /* endif my_rank < TARGET_NTASKS
```
fixed

jedwards4b commented on src/clib/pio_file.c:
```diff
@@ -165,70 +168,83 @@ int PIOc_openfile_retry(const int iosysid, int *ncidp, int *iotype,
-	    file->fh = tmp_fh;
-      
-	*ncidp = file->fh;
+	/* tmp_fh = file->fh; */
```
done

jayeshkrishna commented on src/clib/pio_file.c:
```diff
@@ -2,6 +2,8 @@
 #include <pio.h>
 #include <pio_internal.h>
 
+int pio_next_ncid = 16;
```
Can you add a comment with this info?
